dbms/src/Storages/Transaction/RowCodec.cpp

```diff
@@ -285,53 +290,96 @@ void encodeRowV2(const TiDB::TableInfo & table_info, const std::vector<Field> &
     const ColumnInfos & column_infos,
     ColumnID pk_handle_id,
+    bool fill_column_with_pk_flag,
+    bool force_decode);
+// pre-declar block end
+
+bool appendRowToBlock(
+    const TiKVValue::Base & raw_value,
+    SortedColumnIDWithPosConstIter column_ids_iter,
+    SortedColumnIDWithPosConstIter column_ids_iter_end,
+    Block & block,
+    size_t block_column_pos,
+    const DecodingStorageSchemaSnapshotConstPtr & schema_snapshot,
     bool force_decode)
 {
-    auto row_flag = readLittleEndian<UInt8>(&raw_value[1]);
-    bool is_big = row_flag & RowV2::BigRowMask;
-    return is_big ? appendRowV2ToBlockImpl<true>(raw_value, column_ids_iter, column_ids_iter_end, block, block_column_pos, column_infos, pk_handle_id, force_decode)
-                  : appendRowV2ToBlockImpl<false>(raw_value, column_ids_iter, column_ids_iter_end, block, block_column_pos, column_infos, pk_handle_id, force_decode);
+    const ColumnInfos & column_infos = schema_snapshot->column_infos;
+    // when pk is handle, we need skip pk column when decoding value
+    ColumnID pk_handle_id = InvalidColumnID;
+    if (schema_snapshot->pk_is_handle)
+    {
+        pk_handle_id = schema_snapshot->pk_column_ids[0];
+    }
+
+    // For pk_is_handle table, the column with primary key flag is decoded from encoded key instead of encoded value.
+    // For common handle table, the column with primary key flag is (usually) decoded from encoded key. We skip
+    // filling the columns with primary key flags inside this method.
+    // For other table (non-clustered, use hidden _tidb_rowid as handle), the column with primary key flags could be
+    // changed, we need to fill missing column with default value.
+    const bool fill_missing_column_with_pk_flag = !schema_snapshot->is_common_handle && !schema_snapshot->pk_is_handle;
+
+    switch (static_cast<UInt8>(raw_value[0]))
+    {
+    case static_cast<UInt8>(RowCodecVer::ROW_V2):
+    {
+        auto row_flag = readLittleEndian<UInt8>(&raw_value[1]);
+        bool is_big = row_flag & RowV2::BigRowMask;
+        return is_big ? appendRowV2ToBlockImpl<true>(raw_value, column_ids_iter, column_ids_iter_end, block, block_column_pos, column_infos, pk_handle_id, fill_missing_column_with_pk_flag, force_decode)
+                      : appendRowV2ToBlockImpl<false>(raw_value, column_ids_iter, column_ids_iter_end, block, block_column_pos, column_infos, pk_handle_id, fill_missing_column_with_pk_flag, force_decode);
+    }
+    default:
+        return appendRowV1ToBlock(raw_value, column_ids_iter, column_ids_iter_end, block, block_column_pos, column_infos, pk_handle_id, fill_missing_column_with_pk_flag, force_decode);
+    }
 }
 
-inline bool addDefaultValueToColumnIfPossible(const ColumnInfo & column_info, Block & block, size_t block_column_pos, bool force_decode)
+inline bool addDefaultValueToColumnIfPossible(const ColumnInfo & column_info, Block & block, size_t block_column_pos, bool fill_column_with_pk_flag, bool force_decode)
```

Comment:
I think the variable name "fill_column_with_pk_flag" may by a little difficult to understand.

Reply:
what about `table_is_non_clustered`? The name means that the keys for row data consist of internal `_tidb_rowid` implicitly assigned by TiDB, and we should fill the missing value for those columns with primary key flags under this situation

https://docs.pingcap.com/tidb/dev/clustered-indexes#clustered-indexes

Reply:
How about `ignore_pk_if_absent`?

Reply:
I think `ignore_pk_if_absent` is more straight forward to understand.

Reply:
updated